File: signal-generator/src/models/candle.py
"""Candle data model for OHLCV data."""

from dataclasses import dataclass
from typing import Dict, Any


@dataclass
class Candle:
    """OHLCV candle data."""

    symbol: str
    timestamp: float
    open: float
    high: float
    low: float
    close: float
    volume: float
    timeframe: str = "15"  # minutes

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            'symbol': self.symbol,
            'timestamp': self.timestamp,
            'open': self.open,
            'high': self.high,
            'low': self.low,
            'close': self.close,
            'volume': self.volume,
            'timeframe': self.timeframe
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Candle':
        """Create from dictionary."""
        return cls(**data)

    @property
    def range(self) -> float:
        """Get the candle range (high - low)."""
        return self.high - self.low

    @property
    def body(self) -> float:
        """Get the candle body size (abs(close - open))."""
        return abs(self.close - self.open)

    @property
    def is_bullish(self) -> bool:
        """Check if candle is bullish (close > open)."""
        return self.close > self.open

    @property
    def is_bearish(self) -> bool:
        """Check if candle is bearish (close < open)."""
        return self.close < self.open

File: signal-generator/src/models/candle.js
// Candle data model for OHLCV data
export class Candle {
  constructor(data) {
    this.symbol = data.symbol;
    this.timestamp = data.timestamp;
    this.open = data.open;
    this.high = data.high;
    this.low = data.low;
    this.close = data.close;
    this.volume = data.volume || 0;
    this.timeframe = data.timeframe || '15';
  }

  toDict() {
    return {
      symbol: this.symbol,
      timestamp: this.timestamp,
      open: this.open,
      high: this.high,
      low: this.low,
      close: this.close,
      volume: this.volume,
      timeframe: this.timeframe
    };
  }

  static fromDict(data) {
    return new Candle(data);
  }

  get range() {
    return this.high - this.low;
  }

  get body() {
    return Math.abs(this.close - this.open);
  }

  get isBullish() {
    return this.close > this.open;
  }

  get isBearish() {
    return this.close < this.open;
  }
}

export default Candle;